main.go

```diff
@@ -80,214 +76,232 @@ func main() {
 	glog.Info("Terminated")
 }
 
-// Handles updates for all services and dispatch them between specific per service handlers.
-func handleServices(c *configuration, cloud cloud.Cloud, updates <-chan *registry.ServiceUpdate, done chan struct{}) {
+// Dispatches updates for all services between per service channels.
+func dispatchUpdates(c *configuration, cloud cloud.Cloud, updates <-chan *registry.ServiceUpdate, done chan struct{}) {
 	var wg sync.WaitGroup
-	handlers := make(map[string]chan *registry.ServiceUpdate)
+	serviceUpdatesMap := make(map[string]chan *registry.ServiceUpdate)
 
 	for {
 		select {
 		case update := <-updates:
-			glog.Infof("Handling %s service update", update.ServiceName)
-			// is there a handler for updated service?
-			if handler, ok := handlers[update.ServiceName]; !ok {
+			glog.Infof("Got %q service update", update.ServiceName)
+			// is there a channel for service updates?
+			if serviceUpdates, ok := serviceUpdatesMap[update.ServiceName]; !ok {
 				// no handler
-				handler = make(chan *registry.ServiceUpdate)
-				handlers[update.ServiceName] = handler
+				serviceUpdates = make(chan *registry.ServiceUpdate)
+				serviceUpdatesMap[update.ServiceName] = serviceUpdates
 				// start handler in its own goroutine
 				wg.Add(1)
-				glog.Infof("Initializing a handler for %s service", update.ServiceName)
-				go handleService(c, cloud, update.Tag, handler, &wg, done)
+				glog.Infof("Initializing a handler for %q service updates...", update.ServiceName)
+
+				tagConfig, err := c.GetTagConfiguration(update.Tag)
+				if err != nil {
+					glog.Error(err)
+					continue
+				}
+
+				go handleServiceUpdates(c.Cloud.URLMap, cloud, tagConfig, serviceUpdates, &wg, done)
 			}
-			// send update to handler
-			handlers[update.ServiceName] <- update
+			// send update to channel
+			serviceUpdatesMap[update.ServiceName] <- update
 		case <-done:
-			// wait for all handlers to terminate
+			// wait for all updates are processed
 			wg.Wait()
 			return
 		}
 	}
 }
 
-// handleService handles service updates in a consistent way.
-// It will run until service is deleted or done is closed.
-func handleService(
-	c *configuration,
+// Handles service updates in a consistent way.
+// It's up until service is deleted or done is closed.
+func handleServiceUpdates(
+	urlMapName string,
 	client cloud.Cloud,
-	tag string,
+	tagConfig *tagConfiguration,
 	updates <-chan *registry.ServiceUpdate,
 	wg *sync.WaitGroup,
 	done chan struct{},
 ) {
-	tagInfo, err := parseTag(tag)
-	if err != nil {
-		glog.Error(err)
-		return
-	}
-
-	serviceGroupName := tagInfo.String()
+	tagName := tagConfig.Name
 
-	lock := &sync.RWMutex{}
 	isRunning := false
```

Comment:
How required is this variable? Aren't those operations on the cloud idempotent? Why would we receive these updates on the inconsistent state of the service?

Reply:
Well... looks like the flag is added to prevent handling of hardly possible events in wrong order. I will not change it in the PR since it's not a goal of the PR.